pallets/connectors/src/inbound.rs

```diff
@@ -0,0 +1,270 @@
+		// source destination which should refund the decreased amount.
+		// Blocked by https://github.com/centrifuge/centrifuge-chain/pull/1363
+		// Should be handled by pallet-foreign-investments
+
+		Ok(())
+	}
+
+	/// Increases an existing redemption order of the investor.
+	///
+	/// Transfers the decreased redemption amount from the holdings of the
+	/// `DomainLocator` account of origination domain of this message into the
+	/// investor account.
+	///
+	/// Assumes that the amount of tranche tokens has been locked in the
+	/// `DomainLocator` account of the origination domain beforehand.
+	pub fn do_increase_redemption(
+		pool_id: PoolIdOf<T>,
+		tranche_id: TrancheIdOf<T>,
+		investor: T::AccountId,
+		amount: <T as Config>::Balance,
+		sending_domain: DomainAddress,
+	) -> DispatchResult {
+		// Retrieve investment details
+		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
+
+		// Determine post adjustment amount
+		let pre_amount = T::ForeignInvestment::redemption(&investor, invest_id.clone())?;
+		let post_amount = pre_amount.ensure_add(amount)?;
+
+		// Transfer tranche tokens from `DomainLocator` account of origination domain
+		T::Tokens::transfer(
+			invest_id.clone().into(),
+			&T::AccountConverter::convert(sending_domain),
+			&investor,
+			amount,
+			false,
+		)?;
+
+		T::ForeignInvestment::update_redemption(&investor, invest_id, post_amount)?;
+
+		Ok(())
+	}
+
+	/// Decreases an existing redemption order of the investor.
+	///
+	/// Transfers the decreased redemption amount from the investor account into
+	/// holdings of the `DomainLocator` account of origination domain of this
+	/// message.
+	///
+	/// Initiates a return `ExecutedDecreaseRedemption` message to refund the
+	/// decreased amount on the source domain. The dispatch of this message is
+	/// delayed until the execution of the redemption, e.g. at least until the
+	/// next epoch transition.
+	pub fn do_decrease_redemption(
+		pool_id: PoolIdOf<T>,
+		tranche_id: TrancheIdOf<T>,
+		investor: T::AccountId,
+		currency_index: GeneralCurrencyIndexOf<T>,
+		amount: <T as Config>::Balance,
+		sending_domain: DomainAddress,
+	) -> DispatchResult {
+		// Retrieve investment details
+		let invest_id: T::TrancheCurrency = Self::derive_invest_id(pool_id, tranche_id)?;
+		// NOTE: Required for relaying `ExecutedDecreaseRedemption` message
+		let _currency = Self::try_get_payment_currency(invest_id.clone(), currency_index)?;
+
+		// Determine post adjustment amount
+		let pre_amount = T::ForeignInvestment::redemption(&investor, invest_id.clone())?;
+		let post_amount = pre_amount.ensure_sub(amount)?;
+
+		T::ForeignInvestment::update_redemption(&investor, invest_id.clone(), post_amount)?;
+
+		// Transfer tranche tokens to `DomainLocator` account of
+		// origination domain
+		T::Tokens::transfer(
+			invest_id.into(),
+			&investor,
+			&T::AccountConverter::convert(sending_domain),
+			amount,
+			false,
+		)?;
```

Comment:
I think this should also be handled by the `pallet-foreign-invesments`. WDYT?

Reply:
Sorry, in did not wanted to create more confusion. Given connectors pallet is already huge, I would suggest we create an additional new one

* `pallet-connectors-investments`

This pallet impls `trait Investments` BUT only forwards to an internal type that also implements `trait Investments` BUT does some additional bookkeeping for triggering return messages and minting tokens etc. 

Then we would have 

* `pallet-investments`: Doing the actual investing logic
* `pallet-foreign-investments`: Creating swap orders and async using `pallet-investments` for actual investing
* `pallet-connectors-investments`: Taking care of return messages and conenctors related bookeeping using `pallet-foreign-investments` for actual investing

Does this make sense for you all?

Reply:
IF we agree then:

* `pallet-pool-links` uses `pallet-foreign-investments`
* `pallet-connectors` uses `pallet-connectors-investments`

Reply:
And `pallet-connectors-investments` uses `pallet-foreign-investments`, right?

Reply:
Seems to me like "Taking care of return messages and conenctors related bookeeping" still fits within the intended goal of the `pallet-connectors`, and this only needs more code organization within `pallet-connectors` rather than a separate pallet. But ultimately I don't have a very strong opinion on this and would leave this to you @NunoAlexandre @wischli et al to decide :)

Reply:
I honestly also am not opinionated here. Just be aware that connectors will need to expose some "automation" logic to get start working on investments if they can make progress later down the road.

Reply:
> Taking care of return messages and conenctors related bookeeping" still fits within the intended goal of the pallet-connectors


I agree ☝️

Reply:
Do I understand correctly that the pallet hierarchy for an incoming connectors investment message would be (leaving out gateway) `connectors -> connectors-investments -> foreign-investments -> investments`? E.g.

1. `pallet-connectors` forwards to `pallet-connectors-investments` which basically implements `Investment` as the done in `inbound.rs` in this PR (see associated type `ForeignInvestment`)
2. The `Investment` implementation of `pallet-connectors-investments` forwards the actual investment logic to `pallet-foreign-investments`
3. The `Investment` implementation of `pallet-foreign-investments` creates swap orders and once settled, forwards the investment to the `Investment` implementation of `pallet-investment`

Reply:
> I agree ☝️

I am not completely against it, but we should keep in mind that we will have a lot of async stuff to handle in the future if we allow to change orders on the fly. Just am not so confident that this will not blow the connectors codebase too much. 

But we can go with it of course.